simulation_utils/cram_decontam_func.py

```diff
@@ -0,0 +1,134 @@
+
+from pysam import VariantFile
+from typing import Tuple
+import os
+import logging
+import pickle
+import hailtop.batch as hb
+import hail as hl
+from .generics import *
+
+logging.basicConfig(format="%(levelname)s (%(name)s %(lineno)s): %(message)s")
+logger = logging.getLogger("CHARR simulation pipeline")
+logger.setLevel(logging.INFO)
+
+
+def sample_variant_site(genotype, error, ref, alt1, alt2):
+    p_err = np.random.binomial(1, error)
+    if genotype == (1, 1):
+        allele = random.sample(ALLELES, 1) if p_err else alt1
+    elif genotype == (0, 1):
+        allele = random.sample(ALLELES, 1) if p_err else random.sample((alt1, ref), 1)
+    elif genotype == (1, 2):
+        allele = random.sample(ALLELES, 1) if p_err else random.sample((alt1, alt2), 1)
+    elif genotype == (2, 2):
+        allele = random.sample(ALLELES, 1) if p_err else alt2
+    else:
+        raise Exception("Expecting non reference genotype")
+    return allele
+
+
+def get_var_dict_from_gvcf(input_gvcf: str):
+    # gvcf -> nested dictionary: {chromosome : { position : infomration ...}}
+    # os.system(f"ls {input_gvcf}")
+    print(pysam.__version__)
+    var_dict = {}
+    gvcf_in = VariantFile(input_gvcf)  # auto-detect input format
+    for rec in gvcf_in.fetch():
+        if (rec.samples[0]["GT"][0] != 0) | (rec.samples[0]["GT"][1] != 0):
+            if rec.contig not in var_dict.keys():
+                var_dict[rec.contig] = {}
+            var_dict[rec.contig][rec.pos] = (
+                rec.ref,
+                rec.alts[0],
+                rec.alts[1],
+                rec.samples[0]["GT"],
+            )
+    return var_dict
+
+def write_var_dict_from_gvcf(output: str, gvcf_dict: dict):
+    f = open(f"{output}", "wb")
+    pickle.dump(gvcf_dict, f)
+    f.close()
+
+def read_pickle_dict(dict_path:str):
+    dict = pd.read_pickle(dict_path)
+    return dict
+
+def run_gvcf_dict(b: hb.batch, gvcf_file_name: str, gvcf_path: Tuple):
+    gvcf_dict_path = f"{MY_BUCKET}/gvcf_dicts/{gvcf_file_name}_gvcf_dict.pkl"
+    sample_id = gvcf_file_name.split('.')[0]
+    j = b.new_python_job(name=f"Run_gvcf_dict_{sample_id}")
```

Comment:
You might consider a fixed name, but setting job attributes `j.attributes({'sample_id': sample_id})` i think. You can do this throughout with all the jobs (for the mixing one, do `{'sample': s, 'contaminant': s2}` etc)